### PR TITLE
Make positions fit for meta

### PR DESCRIPTION
**Pull request by odersky**
Simplifications and checks for consistent positions in parse trees.

Review by @liufengyun.

### CODE REVIEW

File: src/dotty/tools/dotc/parsing/JavaParsers.scala [hunk 1]
```diff
@@ -576,35 +584,32 @@ object JavaParsers {
...
       atPos(cdef.pos) {
+        assert(cdef.pos.exists)
         ModuleDef(cdef.name.toTermName,
           makeTemplate(List(), statics, List(), false)).withMods((cdef.mods & (Flags.AccessFlags | Flags.JavaDefined)).toTermFlags)
       }
 
-    private val wild     = Ident(nme.WILDCARD) withPos Position(-1)
-    private val wildList = List(wild) // OPT This list is shared for performance.
+    private def wild     = Ident(nme.WILDCARD)
+    private def wildList = List(wild) // OPT This list is shared for performance.
```
Comment (smarter) [inline src/dotty/tools/dotc/parsing/JavaParsers.scala]: Not shared anymore since it's a def, it could be removed

**Comment by xeno-by** [comment]
@odersky @liufengyun Could you elaborate on the motivation behind this pull request?

**Comment by odersky** [comment]
We want parse trees to be as close to source as possible and we want them to have predictable positions. After all positions is all we have to relate to meta trees.

File: src/dotty/tools/dotc/ast/Desugar.scala [hunk 2]
```diff
@@ -505,7 +505,7 @@ object desugar {
       val clsTmpl = cpy.Template(tmpl)(self = clsSelf, body = tmpl.body)
       val cls = TypeDef(clsName, clsTmpl)
         .withMods(mods.toTypeFlags & RetainedModuleClassFlags | ModuleClassCreationFlags)
-      Thicket(modul, classDef(cls))
+      Thicket(modul, classDef(cls).withPos(mdef.pos))
```
Comment (odersky) [inline src/dotty/tools/dotc/ast/Desugar.scala]: No, the withPos applies to the classDef here, not the thicket.

File: src/dotty/tools/dotc/parsing/Parsers.scala [hunk 3]
```diff
@@ -63,15 +63,16 @@ object Parsers {
       atPos(Position(start, end, point))(t)
 
     def atPos[T <: Positioned](start: Offset, point: Offset)(t: T): T =
-      atPos(start, point, in.lastOffset max start)(t)
+      if (in.lastOffset > start) atPos(start, point, in.lastOffset)(t) else t
```
Comment (odersky) [inline src/dotty/tools/dotc/parsing/Parsers.scala]: No, that's actually a common case. It always happens when the generated tree does not consume any source. In that case we don't set a position and wait for the position to be set by the enclosing node. I'll add a comment to explain.

**Comment by felixmulder** [comment]
@odersky: If we want to merge the syntax highlighting with the scanner/parser we should give comments a token and position as well - not sure if it should be a part of this PR though...

**Comment by odersky** [comment]
@felixmulder Let's discuss comment handling separately.

**Pull request merged**